processor/metric/payload.go

```diff
@@ -0,0 +1,217 @@
+package metric
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/elastic/apm-server/config"
+	m "github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/utility"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+var (
+	processorEntry = common.MapStr{"name": processorName, "event": docType}
+)
+
+type Sample interface {
+	Name() string
+	Value() common.MapStr
+}
+
+type Counter struct {
+	name  string
+	count float64
+	unit  *string
+}
+
+func (c *Counter) Name() string {
+	return c.name
+}
+
+func (c *Counter) Value() common.MapStr {
+	v := common.MapStr{
+		"type":  "counter",
+		"value": c.count,
+	}
+	if c.unit != nil {
+		v["unit"] = c.unit
+	}
+	return v
+}
+
+type Gauge struct {
+	name  string
+	value float64
+	unit  *string
+}
+
+func (g *Gauge) Name() string {
+	return g.name
+}
+
+func (g *Gauge) Value() common.MapStr {
+	v := common.MapStr{
+		"type":  "gauge",
+		"value": g.value,
+	}
+	if g.unit != nil {
+		v["unit"] = g.unit
+	}
+	return v
+}
+
+func CounterFrom(name string, sample map[string]interface{}) *Counter {
```

Comment:
To avoid some repetition you could have a function that returns `(string, float64, *string, error)` and then create a and then the `*From` functions just create `Counter` or `Gauge` out of that.

Reply:
finally refactored this while adding summaries